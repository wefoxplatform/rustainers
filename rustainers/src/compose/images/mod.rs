#![doc = include_str!("./doc.md")]

mod kafka_schema_registry;
pub use self::kafka_schema_registry::*;

mod redpanda;
pub use self::redpanda::*;
