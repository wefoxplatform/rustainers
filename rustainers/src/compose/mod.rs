#![doc = include_str!("./doc.md")]

mod error;
pub use self::error::*;

mod container;
pub use self::container::*;

mod temp_dir;
pub use self::temp_dir::*;

mod service;
pub(crate) use self::service::*;

mod service_state;
pub(crate) use self::service_state::*;

mod options;
pub use self::options::*;

mod runnable;
pub use self::runnable::*;

mod inner;
pub(crate) use self::inner::InnerComposeRunner;

mod runner;

/// Compose Images
pub mod images;
