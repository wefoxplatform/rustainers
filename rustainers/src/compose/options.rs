use std::path::PathBuf;
use std::time::Duration;

use indexmap::IndexMap;
use typed_builder::TypedBuilder;

/// Run options
#[derive(Debug, Clone, TypedBuilder)]
#[builder(doc, field_defaults(default, setter(prefix = "with_")))]
pub struct ComposeRunOption {
    /// Wait interval for service health check
    #[builder(default = Duration::from_secs(1))]
    pub(crate) wait_interval: Duration,

    /// Wait interval for service to exist
    #[builder(default = Duration::from_millis(96))]
    pub(crate) wait_services_interval: Duration,

    /// The environnement variables
    #[builder(setter(transform = |args: impl IntoIterator<Item = (impl Into<String>, impl Into<String>)>| args.into_iter().map(|(k,v)| (k.into(), v.into())).collect()))]
    pub(crate) env: IndexMap<String, String>,

    /// The compose file
    #[builder(setter(strip_option))]
    pub(crate) compose_file: Option<PathBuf>,
    // TODO Wait timeout
}

impl Default for ComposeRunOption {
    fn default() -> Self {
        ComposeRunOption::builder().build()
    }
}
