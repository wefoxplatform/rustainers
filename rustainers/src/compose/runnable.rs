use std::path::Path;

use typed_builder::TypedBuilder;

use super::ComposeService;
use crate::{SharedExposedPort, WaitStrategy};

/// Contains configuration require to run compose containers
#[derive(Debug, Clone, TypedBuilder)]
#[builder(doc, field_defaults(setter(prefix = "with_")))]
#[non_exhaustive]
pub struct RunnableComposeContainers<P> {
    pub(crate) compose_path: P,

    /// The wait condition
    #[builder(default, setter(transform = |args: impl IntoIterator<Item = (impl Into<ComposeService>, impl Into<WaitStrategy>)>| args.into_iter().map(|(k,v)| (k.into(), v.into())).collect()))]
    pub(crate) wait_strategies: Vec<(ComposeService, WaitStrategy)>,

    /// The services port mapping
    #[builder(default, setter(transform = |args: impl IntoIterator<Item = (impl Into<ComposeService>, SharedExposedPort)>| args.into_iter().map(|(k,v)| (k.into(), v)).collect()))]
    pub(crate) port_mappings: Vec<(ComposeService, SharedExposedPort)>,
}

/// Build a runnable compose containers
pub trait ToRunnableComposeContainers {
    /// The path type
    type AsPath: AsRef<Path>;

    /// Should provide the path of the docker-compose file
    ///
    /// # Errors
    ///
    /// If the compose container cannot be created (e.g. issue while creating a [`crate::compose::TemporaryDirectory`])
    fn to_runnable(
        &self,
        builder: RunnableComposeContainersBuilder<Self::AsPath>,
    ) -> RunnableComposeContainers<Self::AsPath>;
}
