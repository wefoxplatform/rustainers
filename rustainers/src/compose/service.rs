use std::convert::Infallible;
use std::fmt::{self, Display};
use std::ops::Index;
use std::str::FromStr;
use std::sync::Arc;

use indexmap::IndexMap;
use tracing::debug;

use crate::ContainerId;

use super::ComposeServiceState;

/// A Compose containers service
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct ComposeService(pub(crate) Arc<str>);

impl Display for ComposeService {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl From<String> for ComposeService {
    fn from(value: String) -> Self {
        Self(Arc::from(value))
    }
}

impl From<&str> for ComposeService {
    fn from(value: &str) -> Self {
        Self(Arc::from(value))
    }
}

impl FromStr for ComposeService {
    type Err = Infallible;

    fn from_str(str: &str) -> Result<Self, Self::Err> {
        Ok(Self(Arc::from(str)))
    }
}

impl AsRef<str> for ComposeService {
    fn as_ref(&self) -> &str {
        &self.0
    }
}

/// A set of services
#[derive(Debug, Clone)]
pub struct Services(pub(crate) IndexMap<ComposeService, ContainerId>);

impl Services {
    /// If it contains the service
    #[must_use]
    pub fn contains(&self, service: &ComposeService) -> bool {
        let result = self.0.contains_key(service);
        debug!(%service, "Service not found");
        result
    }

    /// If it contains all services
    #[must_use]
    pub fn contains_all(&self, services: &[ComposeService]) -> bool {
        services.iter().all(|svc| self.contains(svc))
    }

    /// Get the container id of a service
    pub fn get(&self, service: &ComposeService) -> Option<ContainerId> {
        self.0.get(service).copied()
    }
}

impl Index<&ComposeService> for Services {
    type Output = ContainerId;

    fn index(&self, index: &ComposeService) -> &Self::Output {
        let Some(result) = self.0.get(index) else {
            panic!("Service {index} not found");
        };
        result
    }
}

impl From<Vec<ComposeServiceState>> for Services {
    fn from(value: Vec<ComposeServiceState>) -> Self {
        let map = value
            .into_iter()
            .map(|state| (ComposeService::from(state.service), state.id))
            .collect();
        Self(map)
    }
}
