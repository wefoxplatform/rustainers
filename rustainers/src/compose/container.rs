use std::fmt::{self, Display};
use std::ops::Deref;

use tracing::{error, info};

use crate::compose::ToRunnableComposeContainers;
use crate::runner::Runner;

/// A running compose containers
///
/// If not detached, on drop stop the containers
#[derive(Debug, Clone)]
pub struct ComposeContainers<I>
where
    I: ToRunnableComposeContainers,
{
    pub(crate) runner: Runner,
    pub(crate) name: String,
    pub(crate) images: I,
    pub(crate) file: I::AsPath,
    pub(crate) detached: bool,
}

impl<I> Deref for ComposeContainers<I>
where
    I: ToRunnableComposeContainers,
{
    type Target = I;

    fn deref(&self) -> &Self::Target {
        &self.images
    }
}

impl<I> ComposeContainers<I>
where
    I: ToRunnableComposeContainers,
{
    /// Detach the container
    ///
    /// A detached container won't be stopped during the drop.
    pub fn detach(&mut self) {
        self.detached = true;
    }
}

impl<I> Drop for ComposeContainers<I>
where
    I: ToRunnableComposeContainers,
{
    fn drop(&mut self) {
        let name = &self.name;
        if self.detached {
            info!(%name, "Detached compose containers {self} is NOT stopped");
            return;
        }

        info!(%name, "ðŸš® Stopping compose containers");
        if let Err(e) = self.runner.compose_stop(&self.name, self.file.as_ref()) {
            error!(%name, "Fail to stop compose containers {self} because {e}");
        }
    }
}

impl<I> Display for ComposeContainers<I>
where
    I: ToRunnableComposeContainers,
{
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.name)
    }
}
