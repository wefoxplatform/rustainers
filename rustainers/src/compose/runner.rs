use std::path::Path;

use tracing::{info, warn};

use crate::runner::{Runner, RunnerError};

use super::{
    ComposeContainers, ComposeError, ComposeRunOption, InnerComposeRunner,
    RunnableComposeContainers, ToRunnableComposeContainers,
};

impl Runner {
    /// Run a compose container with the default options
    ///
    /// # Errors
    ///
    /// Fail if the compose containers cannot be started
    pub async fn compose_start<I>(&self, images: I) -> Result<ComposeContainers<I>, RunnerError>
    where
        I: ToRunnableComposeContainers,
    {
        let options = ComposeRunOption::default();
        self.compose_start_with_options(images, options).await
    }

    /// Run compose containers with options
    ///
    /// # Errors
    ///
    /// Fail if the compose containers cannot be started
    pub async fn compose_start_with_options<I>(
        &self,
        images: I,
        options: ComposeRunOption,
    ) -> Result<ComposeContainers<I>, RunnerError>
    where
        I: ToRunnableComposeContainers,
    {
        let mut containers = images.to_runnable(RunnableComposeContainers::builder());
        let file = containers.compose_path.as_ref();
        let wait = &containers.wait_strategies;
        let mappings = &mut containers.port_mappings;

        let name = match self {
            Runner::Docker(runner) => runner.start_compose(file, wait, mappings, options).await,
            Runner::Podman(runner) => runner.start_compose(file, wait, mappings, options).await,
            Runner::Nerdctl(runner) => runner.start_compose(file, wait, mappings, options).await,
        }
        .map_err(|source| RunnerError::ComposeError {
            runner: self.clone(),
            path: file.to_path_buf(),
            source: Box::new(source),
        })?;

        Ok(ComposeContainers {
            runner: self.clone(),
            name,
            images,
            file: containers.compose_path,
            detached: false,
        })
    }

    pub(crate) fn compose_stop(&self, name: &str, file: &Path) -> Result<(), ComposeError> {
        if !file.exists() {
            return Err(ComposeError::ComposeFileMissing(file.to_path_buf()));
        }
        let mut cmd = match self {
            Runner::Docker(runner) => runner.compose_command()?,
            Runner::Podman(runner) => runner.compose_command()?,
            Runner::Nerdctl(runner) => runner.compose_command()?,
        };
        cmd.with_dir(file);
        cmd.push_args(["down"]);
        let status = cmd.status_blocking()?;
        if status.success() {
            info!(%name, "üõë Compose containers stopped");
        } else {
            warn!(%name, ?status, "‚ö†Ô∏è Fail to stop compose containers");
        }
        Ok(())
    }
}

mod docker {
    use std::path::Path;

    use async_trait::async_trait;

    use crate::cmd::Cmd;
    use crate::compose::{ComposeError, ComposeServiceState, InnerComposeRunner, Services};
    use crate::runner::{Docker, InnerRunner};
    use crate::version::Version;

    // https://docs.docker.com/compose/release-notes/#2210
    const PS_JSON_LINES_MINIMAL_VERSION: Version = Version::new(2, 21);

    // See <https://docs.docker.com/compose/release-notes/#2230>
    const NO_TRUNC_MINIMAL_VERSION: Version = Version::new(2, 23);

    #[async_trait]
    impl InnerComposeRunner for Docker {
        fn compose_command(&self) -> Result<Cmd<'static>, ComposeError> {
            if self.compose_version.is_none() {
                return Err(ComposeError::UnsupportedComposeCommand(self.to_string()));
            };
            let mut cmd = self.command();
            cmd.push_arg("compose");
            cmd.ignore_stderr();
            Ok(cmd)
        }

        async fn compose_look_up_services(
            &self,
            _name: &str,
            path: &Path,
        ) -> Result<Services, ComposeError> {
            let mut cmd = self.compose_command()?;
            cmd.with_dir(path);
            let compose_version = self
                .compose_version
                .ok_or(ComposeError::MissingComposeVersion)?;

            let services = if compose_version >= NO_TRUNC_MINIMAL_VERSION {
                cmd.push_args(["ps", "--all", "--no-trunc", "--format", "json"]);
                cmd.json_stream::<ComposeServiceState>().await?
            } else if compose_version >= PS_JSON_LINES_MINIMAL_VERSION {
                cmd.push_args(["ps", "--all", "--format", "json"]);
                cmd.json_stream::<ComposeServiceState>().await?
            } else {
                cmd.push_args(["ps", "--all", "--format", "json"]);
                cmd.json::<Vec<ComposeServiceState>>().await?
            };
            let result = Services::from(services);

            Ok(result)
        }
    }
}

mod nerdctl {
    use async_trait::async_trait;

    use crate::cmd::Cmd;
    use crate::compose::{ComposeError, InnerComposeRunner};
    use crate::runner::{InnerRunner, Nerdctl};

    #[async_trait]
    impl InnerComposeRunner for Nerdctl {
        fn compose_command(&self) -> Result<Cmd<'static>, ComposeError> {
            let mut cmd = self.command();
            cmd.push_arg("compose");
            Ok(cmd)
        }
    }
}

mod podman {
    use std::path::Path;

    use async_trait::async_trait;
    use serde::{Deserialize, Serialize};

    use crate::cmd::Cmd;
    use crate::compose::{ComposeError, ComposeService, InnerComposeRunner, Services};
    use crate::runner::{InnerRunner, Podman};
    use crate::{ContainerHealth, ContainerId, ContainerStatus};

    #[async_trait]
    impl InnerComposeRunner for Podman {
        fn compose_command(&self) -> Result<Cmd<'static>, ComposeError> {
            if self.compose_version.is_none() {
                return Err(ComposeError::UnsupportedComposeCommand(self.to_string()));
            };
            let mut cmd = Cmd::new("podman-compose");
            cmd.ignore_stderr();
            Ok(cmd)
        }

        async fn compose_look_up_services(
            &self,
            name: &str,
            _path: &Path,
        ) -> Result<Services, ComposeError> {
            // To use the JSON output, we need to use the standard ps command of podman
            let mut cmd = self.command();
            let label = format!(
                "label=io.podman.compose.project={}",
                name.to_ascii_lowercase()
            );
            cmd.push_args(["ps", "--all", "--filter", &label, "--format", "json"]);
            let containers = cmd.json::<Vec<PodmanComposeServiceState>>().await?;
            let result = containers
                .into_iter()
                .map(|it| (ComposeService::from(it.labels.service), it.id))
                .collect();
            Ok(Services(result))
        }
    }

    #[derive(Debug, Clone, Serialize, Deserialize)]
    #[serde(rename_all = "PascalCase")]
    pub struct PodmanComposeServiceState {
        pub(crate) id: ContainerId,
        names: Vec<String>,
        labels: PodmanComposeLabels,
        state: ContainerStatus,
        health: Option<ContainerHealth>,
        exit_code: Option<i32>,
    }

    #[derive(Debug, Clone, Serialize, Deserialize)]
    pub struct PodmanComposeLabels {
        #[serde(rename = "com.docker.compose.container-number")]
        container_number: String,
        #[serde(rename = "com.docker.compose.project")]
        project: String,
        #[serde(rename = "com.docker.compose.project.config_files")]
        config_files: String,
        #[serde(rename = "com.docker.compose.project.working_dir")]
        working_dir: String,
        #[serde(rename = "com.docker.compose.service")]
        service: String,
    }

    #[cfg(test)]
    mod tests {
        use super::*;

        #[test]
        fn should_serde_podman_service() {
            let json = include_str!("../../tests/assets/podman_lookup.json");
            let services =
                serde_json::from_str::<Vec<PodmanComposeServiceState>>(json).expect("json");
            insta::assert_json_snapshot!(services);
        }
    }
}
