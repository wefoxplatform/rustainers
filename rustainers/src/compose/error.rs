use std::path::PathBuf;

use crate::runner::ContainerError;

use super::ComposeService;

/// A compose error
#[derive(Debug, thiserror::Error)]
#[non_exhaustive]
pub enum ComposeError {
    /// Unsupported compose command
    #[error("Compose command is unsupported for {0}")]
    UnsupportedComposeCommand(String),

    /// Compose file missing
    #[error("Compose file missing {0:?}")]
    ComposeFileMissing(PathBuf),

    /// Bad compose file missing
    #[error("Bad compose file {0:?}")]
    BadComposeFile(PathBuf),

    /// Cannot launch compose containers
    #[error("Cannot launch compose containers {0:?}")]
    ComposeContainerCannotBeStarted(String),

    /// Custom health forbidden in compose
    #[error("Cannot use a custom health check with compose service {0}")]
    NoCustomHealthCheckInCompose(ComposeService),

    /// A temporary directory error
    #[error(transparent)]
    TempDirError(#[from] TempDirError),

    /// A container error
    #[error(transparent)]
    ContainerError(#[from] ContainerError),

    /// Fail to run command error
    #[error(transparent)]
    CommandError(#[from] crate::cmd::CommandError),

    /// Cannot extract compose service state
    #[error("Cannot extract compose service state because {source}\njson: {json}")]
    CannotParseComposeServiceState {
        /// The json
        json: String,
        /// The source
        source: serde_json::Error,
    },

    /// Missing compose version
    #[error("Missing compose version")]
    MissingComposeVersion,
}

/// A temporary directory error
#[derive(Debug, thiserror::Error)]
#[non_exhaustive]
pub enum TempDirError {
    /// Cannot create a temp. file from absolute path
    #[error("Cannot create a temporary file from absolute path {0:?}")]
    CannotCreateAbsoluteTempFile(PathBuf),

    /// Cannot override a temp. file from absolute path
    #[error("Cannot override a temporary file from absolute path {0:?}")]
    CannotOverrideTempFile(PathBuf),

    /// Cannot create directory
    #[error("Cannot create dir {dir:?} because {source}")]
    CannotCreateDir {
        /// The dir to create
        dir: PathBuf,
        /// The source
        source: std::io::Error,
    },

    /// Cannot write file
    #[error("Cannot write {file:?} because {source}")]
    CannotWriteFile {
        /// The file to write
        file: PathBuf,
        /// The source
        source: std::io::Error,
    },

    /// Cannot set permission
    #[error("Cannot write {file:?} because {source}")]
    CannotSetPermission {
        /// The file to write
        file: PathBuf,
        /// The source
        source: std::io::Error,
    },
}
