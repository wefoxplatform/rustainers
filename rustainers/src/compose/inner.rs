use std::ffi::OsStr;
use std::path::Path;
use std::time::Duration;

use async_trait::async_trait;
use tracing::info;

use crate::cmd::Cmd;
use crate::runner::InnerRunner;
use crate::{ContainerId, ExposedPort, WaitStrategy};

use super::{ComposeError, ComposeRunOption, ComposeService, ComposeServiceState, Services};

#[async_trait]
pub(crate) trait InnerComposeRunner: InnerRunner {
    fn compose_command(&self) -> Result<Cmd<'static>, ComposeError>;

    #[tracing::instrument(skip(self), fields(runner = %self))]
    async fn start_compose(
        &self,
        dir: &Path,
        wait_strategies: &[(ComposeService, WaitStrategy)],
        port_mappings: &mut [(ComposeService, ExposedPort)],
        options: ComposeRunOption,
    ) -> Result<String, ComposeError> {
        let Some(name) = dir.file_name().and_then(OsStr::to_str).map(str::to_string) else {
            return Err(ComposeError::BadComposeFile(dir.to_path_buf()))?;
        };
        self.compose_up(&name, dir, &options).await?;

        // Find required services
        let required_services = wait_strategies
            .iter()
            .map(|(svc, _)| svc.clone())
            .collect::<Vec<_>>();
        let services = self
            .find_required_services(
                &name,
                &required_services,
                options.wait_services_interval,
                dir,
            )
            .await?;

        // Wait
        let interval = options.wait_interval;
        for (service, wait) in wait_strategies {
            debug_assert!(services.contains(service));
            #[allow(clippy::indexing_slicing)]
            let id = services[service];
            self.wait_service_ready(service, id, wait, interval).await?;
        }

        // Port mapping
        for (service, mapping) in port_mappings {
            debug_assert!(services.contains(service));
            #[allow(clippy::indexing_slicing)]
            let id = services[service];
            let port = self.port(id, mapping.container_port).await?;
            mapping.bind_port(port).await;
        }

        Ok(name)
    }

    #[tracing::instrument(level = "debug", skip(self), fields(runner = %self))]
    async fn wait_service_ready(
        &self,
        service: &ComposeService,
        id: ContainerId,
        wait_condition: &WaitStrategy,
        interval: Duration,
    ) -> Result<(), ComposeError> {
        self.wait_ready(id, wait_condition, interval).await?;
        Ok(())
    }

    #[tracing::instrument(level = "debug", skip(self), fields(runner = %self))]
    async fn compose_up(
        &self,
        name: &str,
        dir: &Path,
        options: &ComposeRunOption,
    ) -> Result<(), ComposeError> {
        info!(%name, ?dir, "ðŸš€ Launching compose container");
        let mut cmd = self.compose_command()?;
        cmd.with_dir(dir);
        cmd.push_args(["up", "--detach"]);
        if let Some(file) = options.compose_file.as_ref().and_then(|it| it.to_str()) {
            cmd.push_args(["--file", file]);
        }
        cmd.set_env(options.env.clone());

        let cmd_err = cmd.clone();
        let status = cmd.status().await?;
        if status.success() {
            Ok(())
        } else {
            Err(ComposeError::ComposeContainerCannotBeStarted(
                cmd_err.to_string(),
            ))
        }
    }

    async fn compose_look_up_services(
        &self,
        _name: &str,
        path: &Path,
    ) -> Result<Services, ComposeError> {
        let mut cmd = self.compose_command()?;
        cmd.with_dir(path);
        cmd.push_args(["ps", "--all", "--no-trunc", "--format", "json"]);
        let states = cmd.json_stream::<ComposeServiceState>().await?;
        let result = Services::from(states);
        Ok(result)
    }

    #[tracing::instrument(skip(self), fields(runner = %self))]
    async fn find_required_services(
        &self,
        name: &str,
        required_services: &[ComposeService],
        interval: Duration,
        path: &Path,
    ) -> Result<Services, ComposeError> {
        loop {
            let result = self.compose_look_up_services(name, path).await?;
            if result.contains_all(required_services) {
                return Ok(result);
            }
            tokio::time::sleep(interval).await;
        }
    }
}
