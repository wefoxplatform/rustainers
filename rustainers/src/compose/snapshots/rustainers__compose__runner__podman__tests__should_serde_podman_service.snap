---
source: rustainers/src/compose/runner.rs
expression: services
---
[
  {
    "Id": "2307493ebc7c2cf221e339ecc23c0d7c53065f0e01bc374cf1af13876062aee3",
    "Names": [
      "tc-kafka"
    ],
    "Labels": {
      "com.docker.compose.container-number": "1",
      "com.docker.compose.project": "kafka_schema_registry",
      "com.docker.compose.project.config_files": "docker-compose.yaml",
      "com.docker.compose.project.working_dir": "/Users/igor.laborie/Documents/Workspaces/Wefox/shared/wefox-ai-wai-rust/wai-testcontainers/src/compose/kafka_schema_registry",
      "com.docker.compose.service": "kafka"
    },
    "State": "running",
    "Health": null,
    "ExitCode": 0
  },
  {
    "Id": "e383b5e5d80e2f7ab640940e14221a66d07914cad2d1189076efc43ea03609b7",
    "Names": [
      "tc-schema-registry"
    ],
    "Labels": {
      "com.docker.compose.container-number": "1",
      "com.docker.compose.project": "kafka_schema_registry",
      "com.docker.compose.project.config_files": "docker-compose.yaml",
      "com.docker.compose.project.working_dir": "/Users/igor.laborie/Documents/Workspaces/Wefox/shared/wefox-ai-wai-rust/wai-testcontainers/src/compose/kafka_schema_registry",
      "com.docker.compose.service": "schema-registry"
    },
    "State": "running",
    "Health": null,
    "ExitCode": 0
  },
  {
    "Id": "93b68542fcc252e3734c4d28a028c41240915b062b4ba1fc804e960486540c79",
    "Names": [
      "tc-kafdrop"
    ],
    "Labels": {
      "com.docker.compose.container-number": "1",
      "com.docker.compose.project": "kafka_schema_registry",
      "com.docker.compose.project.config_files": "docker-compose.yaml",
      "com.docker.compose.project.working_dir": "/Users/igor.laborie/Documents/Workspaces/Wefox/shared/wefox-ai-wai-rust/wai-testcontainers/src/compose/kafka_schema_registry",
      "com.docker.compose.service": "kafdrop"
    },
    "State": "running",
    "Health": null,
    "ExitCode": 0
  }
]
